File: app/src/main/java/calendario/kevshupp/diariokevinali/AlbumFragment.java
package calendario.kevshupp.diariokevinali;

import android.graphics.Color;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.PopupMenu;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.QueryDocumentSnapshot;

import java.util.ArrayList;
import java.util.List;

public class AlbumFragment extends Fragment {

    private String coupleId, userId, userName, userImageUri, theme;
    private FirebaseFirestore db;
    private AlbumGalleryAdapter adapter;
    private List<Message> moments = new ArrayList<>();
    private AlbumManager albumManager;

    public static AlbumFragment newInstance(String coupleId, String userId, String userName, String userImageUri, String theme) {
        AlbumFragment fragment = new AlbumFragment();
        Bundle args = new Bundle();
        args.putString("coupleId", coupleId);
        args.putString("userId", userId);
        args.putString("userName", userName);
        args.putString("userImageUri", userImageUri);
        args.putString("theme", theme);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            coupleId = getArguments().getString("coupleId");
            userId = getArguments().getString("userId");
            userName = getArguments().getString("userName");
            userImageUri = getArguments().getString("userImageUri");
            theme = getArguments().getString("theme");
        }
        db = FirebaseFirestore.getInstance();
        albumManager = new AlbumManager(requireContext(), coupleId, userId, userName, userImageUri);
        albumManager.setTheme(theme);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.dialog_shared_album, container, false);

        if ("Pixel Oscuro".equals(theme)) {
            view.setBackgroundResource(R.drawable.bg_parchment_pixel_dark);
            ((TextView) view.findViewById(R.id.tvAlbumTitle)).setTextColor(Color.WHITE);
            ((TextView) view.findViewById(R.id.tvAlbumSubtitle)).setTextColor(Color.LTGRAY);
            view.findViewById(R.id.llAlbumHeader).setBackgroundResource(R.drawable.bg_parchment_pixel_dark);
            Button btnAddMoment = view.findViewById(R.id.btnAddMoment);
            btnAddMoment.setTextColor(Color.WHITE);
            btnAddMoment.setBackgroundTintList(android.content.res.ColorStateList.valueOf(Color.parseColor("#1A1A2E")));
        } else {
            view.setBackgroundResource(R.drawable.bg_parchment_pixel);
        }

        Button btnAdd = view.findViewById(R.id.btnAddMoment);
        btnAdd.setOnClickListener(v -> albumManager.showAddMomentDialog(new AlbumManager.AlbumCallback() {
            @Override public void onPickImage() { 
                if (getActivity() instanceof MainActivity) ((MainActivity) getActivity()).pickImage(4); 
            }
            @Override public void onMomentSaved() {
                if (getActivity() instanceof MainActivity) ((MainActivity) getActivity()).sendNotificationV1("Nuevo momento en el álbum 📸");
            }
        }));

        RecyclerView rv = view.findViewById(R.id.rvAlbumPhotos);
        adapter = new AlbumGalleryAdapter(moments, new AlbumGalleryAdapter.OnMomentClickListener() {
            @Override public void onMomentClick(Message m) {
                albumManager.showAlbumDetail(m);
            }

            @Override public void onMomentLongClick(View v, Message m) {
                if (m.getAuthorId().equals(userId)) {
                    PopupMenu p = new PopupMenu(getContext(), v);
                    p.getMenu().add("Editar Momento");
                    p.getMenu().add("Eliminar momento completo");
                    p.setOnMenuItemClickListener(menuItem -> {
                        String title = menuItem.getTitle().toString();
                        if (title.equals("Editar Momento")) {
                            albumManager.showEditAlbumDialog(m);
                        } else if (title.equals("Eliminar momento completo")) {
                            db.collection("messages").document(m.getMessageId()).delete();
                        }
                        return true;
                    });
                    p.show();
                }
            }
        });
        adapter.setTheme(theme);
        rv.setLayoutManager(new GridLayoutManager(getContext(), 4));
        rv.setAdapter(adapter);

        listenAlbum();

        return view;
    }

    private void listenAlbum() {
        db.collection("messages").whereEqualTo("partnerId", coupleId)
            .addSnapshotListener((shots, error) -> {
                if (shots != null) {
                    moments.clear();
                    for (QueryDocumentSnapshot doc : shots) {
                        Message m = doc.toObject(Message.class);
                        if (m.getContent() != null && m.getContent().startsWith("[ALBUM]")) moments.add(m);
                    }
                    moments.sort((m1, m2) -> Long.compare(m2.getTimestamp(), m1.getTimestamp()));
                    adapter.notifyDataSetChanged();
                }
            });
    }

    public void addImageUrl(String url) {
        if (albumManager != null) albumManager.addImageUrl(url);
    }
}

File: app/src/main/java/calendario/kevshupp/diariokevinali/AlbumFragment.kt
package calendario.kevshupp.diariokevinali

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.platform.ComposeView
import androidx.fragment.app.Fragment
import calendario.kevshupp.diariokevinali.compose.AlbumScreen
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.ListenerRegistration
import com.google.firebase.firestore.Query
import kotlinx.coroutines.withContext
import kotlinx.coroutines.Dispatchers

class AlbumFragment : Fragment() {
    private var coupleId: String? = null
    private var userId: String? = null
    private var userName: String? = null
    private var userImageUri: String? = null
    private var theme: String = "Pixel Claro"
    private val db = FirebaseFirestore.getInstance()
    private var albumListener: ListenerRegistration? = null
    private var albumManager: AlbumManager? = null

    companion object {
        @JvmStatic
        fun newInstance(coupleId: String, userId: String, userName: String, userImageUri: String?, theme: String): AlbumFragment {
            val fragment = AlbumFragment()
            val args = Bundle()
            args.putString("coupleId", coupleId)
            args.putString("userId", userId)
            args.putString("userName", userName)
            args.putString("userImageUri", userImageUri)
            args.putString("theme", theme)
            fragment.arguments = args
            return fragment
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (arguments != null) {
            coupleId = arguments?.getString("coupleId")
            userId = arguments?.getString("userId")
            userName = arguments?.getString("userName")
            userImageUri = arguments?.getString("userImageUri")
            theme = arguments?.getString("theme") ?: "Pixel Claro"
        }
        val safeCoupleId = coupleId ?: ""
        val safeUserId = userId ?: ""
        val safeUserName = userName ?: ""
        albumManager = AlbumManager(requireContext(), safeCoupleId, safeUserId, safeUserName, userImageUri)
        albumManager?.setTheme(theme)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        return ComposeView(requireContext()).apply {
            setContent {
                var moments by remember { mutableStateOf(listOf<Message>()) }
                var localPhotos by remember { mutableStateOf(listOf<LocalPhoto>()) }
                var localFolderConfigured by remember { mutableStateOf(false) }
                var isLocalPhotosLoading by remember { mutableStateOf(false) }
                var localPhotosReloadTrigger by remember { mutableStateOf(0) }
                val isDark = theme == "Pixel Oscuro"
                val activityRef = activity as? MainActivity

                LaunchedEffect(Unit) {
                    if (coupleId.isNullOrEmpty()) {
                        val prefs = activityRef?.getSharedPreferences("DiarioPrefs", android.content.Context.MODE_PRIVATE)
                        coupleId = prefs?.getString("coupleId", coupleId)
                        userId = prefs?.getString("userId", userId)
                        userName = prefs?.getString("userName", userName)
                        userImageUri = prefs?.getString("userImage", userImageUri)
                        theme = prefs?.getString("theme", theme) ?: theme
                        albumManager?.setTheme(theme)
                    }
                }

                LaunchedEffect(coupleId, localPhotosReloadTrigger) {
                    val prefs = activityRef?.getSharedPreferences("DiarioPrefs", android.content.Context.MODE_PRIVATE)
                    val folderUri = prefs?.getString("syncLocalFolderUri", null)
                    val configured = !folderUri.isNullOrEmpty()
                    localFolderConfigured = configured
                    if (configured) {
                        isLocalPhotosLoading = true
                        withContext(Dispatchers.IO) {
                            albumManager?.getLocalPhotos() ?: emptyList()
                        }.let { photos ->
                            localPhotos = photos
                            isLocalPhotosLoading = false
                        }
                    }
                }

                DisposableEffect(coupleId) {
                    albumListener?.remove()
                    albumListener = coupleId?.let { id ->
                        db.collection("messages")
                            .whereEqualTo("partnerId", id)
                            .orderBy("timestamp", Query.Direction.DESCENDING)
                            .addSnapshotListener { shots, _ ->
                                if (shots != null) {
                                    val items = shots.mapNotNull { doc ->
                                        val m = doc.toObject(Message::class.java)
                                        if (m.content?.startsWith("[ALBUM]") == true) m else null
                                    }
                                    moments = items
                                }
                            }
                    }
                    onDispose {
                        albumListener?.remove()
                        albumListener = null
                    }
                }

                AlbumScreen(
                    moments = moments,
                    localPhotos = localPhotos,
                    localFolderConfigured = localFolderConfigured,
                    isLocalPhotosLoading = isLocalPhotosLoading,
                    onReloadLocalPhotos = { localPhotosReloadTrigger++ },
                    theme = theme,
                    onAddMoment = {
                        albumManager?.showAddMomentDialog(object : AlbumManager.AlbumCallback {
                            override fun onPickImage() {
                                activityRef?.pickImage(4)
                            }

                            override fun onMomentSaved() {
                                activityRef?.updatePetOnInteraction()
                                activityRef?.sendNotificationV1("¡Nuevo recuerdo en el álbum! 📸", "${userName ?: ""} agregó un momento al álbum", null, "album")
                            }
                        })
                    },
                    onOpenMoment = { msg ->
                        albumManager?.showAlbumDetail(msg)
                    },
                    onEditMoment = { msg ->
                        albumManager?.showEditAlbumDialog(msg)
                    },
                    onDeleteMoment = { msg ->
                        msg.messageId?.let { db.collection("messages").document(it).delete() }
                    },
                    onOpenLocalPhoto = { url ->
                        albumManager?.showFullScreenImage(url, null) {
                            localPhotosReloadTrigger++
                        }
                    },
                    isOwner = { msg -> msg.authorId == userId }
                )
            }
        }
    }
}
